src/haven.rs

```diff
@@ -153,3 +178,46 @@ pub async fn haven_loop(ctx: DaemonContext, haven_cfg: HavenForwardConfig) -> an
             .await?;
     }
 }
+
+async fn tcp_forward(ctx: DaemonContext, haven_cfg: HavenForwardConfig) -> anyhow::Result<()> {
+    async fn stream_loop(earendil_stream: Stream, tcp_stream: TcpStream) -> anyhow::Result<()> {
```

Comment:
I actually think using functions in functions like this is not super useful unless you're using a closure to capture stuff. (I know we do it several places but I will probably refactor things out)

Just move it out; if the scope has too many random stuff that's an indication of making separate files/modules, and not making things local and making functions long